framework/aster-frame/src/task/scheduler.rs

```diff
@@ -1,64 +1,146 @@
 // SPDX-License-Identifier: MPL-2.0
 
+use crate::config::SCHED_DEBUG_LOG;
 use crate::prelude::*;
-use crate::sync::SpinLock;
-use crate::task::Task;
+use crate::sync::{SpinLock, SpinLockGuard};
+use crate::task::{NeedResched, ReadPriority, Task};
 
-use lazy_static::lazy_static;
+use log::debug;
 
-lazy_static! {
-    pub(crate) static ref GLOBAL_SCHEDULER: SpinLock<GlobalScheduler> =
-        SpinLock::new(GlobalScheduler { scheduler: None });
+pub(crate) static GLOBAL_SCHEDULER: spin::Once<SpinLock<GlobalScheduler>> = spin::Once::new();
+
+pub fn init() {
+    GLOBAL_SCHEDULER.call_once(|| SpinLock::new(GlobalScheduler { scheduler: None }));
 }
 
 /// A scheduler for tasks.
 ///
-/// An implementation of scheduler can attach scheduler-related information
-/// with the `TypeMap` returned from `task.data()`.
-pub trait Scheduler: Sync + Send {
-    fn enqueue(&self, task: Arc<Task>);
+/// Operations on the scheduler should be performed with interrupts disabled,
+/// which has been ensured by the callers of the `GLOBAL_SCHEDULER`.
+/// Therefore, implementations of this trait do not need to worry about interrupt safety.
+pub trait Scheduler<T: NeedResched + ReadPriority = Task>: Sync + Send {
+    /// Called when a task enters a runnable state.
+    /// Add the task to the scheduler.
+    fn enqueue(&self, task: Arc<T>);
+
+    /// Called when the task is no longer alive.
+    /// Remove the task-related from the scheduler.
+    ///
+    /// A similar method in the linux scheduler is `dequeue_task`.
+    ///
+    /// Return `true` if the task was in the scheduler.
+    fn remove(&self, task: &Arc<T>) -> bool;
+
+    /// Choose the most appropriate task eligible to run next.
+    fn pick_next_task(&self) -> Option<Arc<T>>;
 
-    fn dequeue(&self) -> Option<Arc<Task>>;
+    /// Tells whether the given current task should be preempted
+    /// by other tasks in the queue.
+    fn should_preempt(&self, cur_task: &Arc<T>) -> bool;
```

Comment:
Why should `cur_task` be a parameter, instead of just using `Task::current` to get current task?

```rust
fn should_preempt(&self) -> bool {
    let cur_task = Task::current();
    // Original logic
}

Reply:
Options:
1. remove `cur_task`-like params from `Scheduler` trait, use `Task::current()` instead;
2. let the users of `Scheduler` implementations to explicitly pass the current task(in type `Arc<T>`) held by the processor.

Option 1 makes better guarantees on the correctness of `cur_task`, but may result in dead locks when being called after the processor is locked. And a new trait may be needed to decouple type `Task` and the sched impls.
The correctness of `cur_task` in option 2 can only depends on developers. However, it isolates the `Scheduler` impls from the processor impl & type `Task`, and thus provides a little more feasibility for testing.

Reply:
Or, we can use `Task::current()` in `GlobalScheduler`? Currently, the struct is kind of tedious.

```rust
impl GlobalScheduler {
    pub fn should_preempt_cur_task(&self) -> bool {
        let cur_task = Task::current();
        self.scheduler.unwrap().should_preempt(cur_task)
    }
    // ...
```

Reply:
I believe the issue revolves around whether any task, other than the currently running task, can be preempted. If only preemption to current task is allowed, the scheduler should not provide an interface that suggests any tasks can be preempted.

> but may result in dead locks when being called after the processor is locked

It is surprising. Since `PROCESSOR` implies a per-CPU variable, there should be no need to protect it with a lock, correct? However, the problem of potential deadlocks should be addressed separately and should not be directly related to the interface.

> And a new trait may be needed to decouple type Task and the sched impls.

The current `Scheduler` accepts generic paramter which implements `NeedResched + ReadPriority`, how about merge these traits as one, for example, `SchedEntity`?
```rust
pub trait SchedEntity {
    // The methods of NeedResched and ReadPriority
}
```

Reply:
`cur_task`-like params in `Scheduler` trait are removed.

Trait `Current` is added to offer a `current` method for the generic task type.
`SchedTaskBase` inherits the three traits below to claim a basic requirement of the scheduled tasks.
Trait `WakeUp` is not added in `SchedTaskBase` because it is only used by O(1) scheduler till now. Thus, other developers does not need to implement the trait for a mock task type when writing unit tests for their schedulers.

```rust
pub trait SchedTaskBase: ReadPriority + NeedResched + Current {}
```